lib/watir/locators/element/selector_builder/xpath.rb

```diff
@@ -3,32 +3,67 @@ module Locators
     class Element
       class SelectorBuilder
         class XPath
-          def initialize(should_use_label_element)
-            @should_use_label_element = should_use_label_element
-          end
-
-          def build(selectors)
-            adjacent = selectors.delete :adjacent
-            xpath = adjacent ? process_adjacent(adjacent) : './/*'
+          # Regular expressions that can be reliably converted to xpath `contains`
+          # expressions in order to optimize the locator.
+          CONVERTABLE_REGEXP = /
+          \A
+            ([^\[\]\\^$.|?*+()]*) # leading literal characters
+            [^|]*?                # do not try to convert expressions with alternates
+            ([^\[\]\\^$.|?*+()]*) # trailing literal characters
+          \z
+        /x
 
-            tag_name = selectors.delete(:tag_name).to_s
-            xpath << "[local-name()='#{tag_name}']" unless tag_name.empty?
+          def initialize(use_label_element)
+            @should_use_label_element = use_label_element
+          end
 
-            index = selectors.delete(:index)
+          def build(selector, values_to_match)
+            adjacent = selector.delete :adjacent
+            xpath = adjacent.nil? ? default_start : process_adjacent(adjacent)
+            xpath << add_tag_name(selector)
+            index = selector.delete(:index)
 
             # the remaining entries should be attributes
-            xpath << '[' << attribute_expression(nil, selectors) << ']' unless selectors.empty?
+            xpath << add_attributes(selector)
 
             xpath << "[#{index + 1}]" if adjacent && index
 
-            p xpath: xpath, selectors: selectors if $DEBUG
+            xpath = add_regexp_predicates(xpath, values_to_match)
+
+            Watir.logger.debug(xpath: xpath, selector: selector)
 
             [:xpath, xpath]
           end
 
+          def default_start
```

Comment:
Can it be a constant?

Reply:
the point of the method is that this value is different depending on the subclass, which constants are not good for.